lib/logstash/pluginmanager/install.rb

```diff
@@ -1,82 +1,128 @@
 
-    puts ("validating #{plugin} #{version}")
+    # install_list will be an array of [plugin name, version] tuples, version can be nil
+    install_list = []
 
-    unless gem_path = (plugin =~ /\.gem$/ && File.file?(plugin)) ? plugin : LogStash::PluginManager::Util.download_gem(plugin, version)
-      $stderr.puts ("Plugin does not exist '#{plugin}'. Aborting")
-      return 99
-    end
+    if development?
+      specs = LogStash::PluginManager.all_installed_plugins_gem_specs(gemfile)
+      install_list = specs.inject([]) do |result, spec|
+        result = result + spec.dependencies.select{|dep| dep.type == :development}.map{|dep| [dep.name] + dep.requirement.as_list + [{:group => :development}]}
+      end
+    else
+      # at this point we know that plugin_list is not empty and if the --version is specified there is only one plugin in plugin_list
+
+      install_list = version ? [plugin_list << version] : plugin_list.map{|plugin| [plugin, nil]}
 
-    unless gem_meta = LogStash::PluginManager::Util.logstash_plugin?(gem_path)
-      $stderr.puts ("Invalid logstash plugin gem '#{plugin}'. Aborting...")
-      return 99
+      install_list.each do |tuple|
+        puts("Validating #{tuple.compact.join("-")}")
+        unless LogStash::PluginManager.is_logstash_plugin?(*tuple)
+          $stderr.puts("Installation aborted")
+          exit(1)
+        end
+      end unless force?
+
+      # at this point we know that we either have a valid gem name & version or a valid .gem file path
+
+      # if LogStash::PluginManager.is_plugin_file?(plugin)
+      #   exit(1) unless cache_gem_file(plugin)
+      #   spec = LogStash::PluginManager.plugin_file_spec(plugin)
+      #   gemfile.update(spec.name, spec.version.to_s)
+      # else
+      #   plugins.each{|tuple| gemfile.update(*tuple)}
+      # end
     end
 
-    puts ("Valid logstash plugin. Continuing...")
 
-    if LogStash::PluginManager::Util.installed?(gem_meta.name)
+    install_list = LogStash::PluginManager.merge_duplicates(install_list)
+    install_list.each{|tuple| gemfile.update(*tuple)}
+    gemfile.save
 
-      current = Gem::Specification.find_by_name(gem_meta.name)
-      if Gem::Version.new(current.version) > Gem::Version.new(gem_meta.version)
-        unless LogStash::PluginManager::Util.ask_yesno("Do you wish to downgrade this plugin?")
-          $stderr.puts("Aborting installation")
-          return 99
-        end
-      end
+    puts("Installing" + (install_list.empty? ? "..." : " " + install_list.map{|tuple| tuple.first}.join(", ")))
 
-      puts ("removing existing plugin before installation")
-      ::Gem.done_installing_hooks.clear
-      ::Gem::Uninstaller.new(gem_meta.name, {:force => true}).uninstall
+    bundler_options = {:install => true}
+    bundler_options = bundler_options.merge({:without => []}) if development?
+
+    # any errors will be logged to $stderr by invoke_bundler!
+    output, exception = LogStash::Bundler.invoke_bundler!(bundler_options)
```

Comment:
granted this is strange but this is necessary here because of the stdout capturing... in particular upon error where you still want to keep and return the captured IO.